src/routes/calendar.rs

```diff
@@ -25,61 +24,185 @@ pub struct Search {
+    let (jar, selected) = if course_exists {
+        // no-op if course is already in state
+        (CookieJar::new(), selected)
+    } else {
+        let default_sections = state.default_thin_sections(&term, course.clone())?;
 
-    debug!(?default_sections);
+        let mut new_selected = selected.clone();
+        new_selected.courses.insert(course, default_sections);
 
-    let jar = CookieJar::new().add({
-        let mut user_state = selected_courses.clone();
-        user_state.courses.insert(course, default_sections);
+        (
+            CookieJar::new().add(new_selected.make_cookie(term)),
+            new_selected,
+        )
+    };
 
-        user_state.make_cookie(term)
-    });
+    let courses = state.courses(term, &selected.thin_courses())?;
 
     Ok((
         jar,
         html! {
             (calendar_view_container(true))
-            (courses_container(true))
+            (courses_container(true, term, &courses, &selected))
         },
     ))
 }
 
-#[instrument(level = "debug", skip(_state))]
+#[instrument(level = "debug", skip(state))]
 pub async fn rm_from_calendar(
     Path(term): Path<Term>,
-    State(_state): State<Arc<DatabaseAppState>>,
-    selected_courses: SelectedCourses,
-    Form(Search { course }): Form<Search>,
+    State(state): State<Arc<DatabaseAppState>>,
+    selected: SelectedCourses,
+    Query(Search { course }): Query<Search>,
 ) -> Result<impl IntoResponse, AppError> {
     // no-op if course is not in cookie
-    if !selected_courses.courses.keys().any(|c| *c == course) {
-        return Ok((CookieJar::new(), html!()));
+    if !selected.courses.keys().any(|c| *c == course) {
+        let courses = state.courses(term, &selected.thin_courses())?;
+
+        return Ok((
+            CookieJar::new(),
+            html! {
+                (calendar_view_container(true))
+                (courses_container(true, term, &courses, &selected))
+            },
+        ));
     }
 
-    let jar = CookieJar::new().add({
-        let mut user_state = selected_courses;
-        user_state
-            .courses
-            .retain(|thin_course, _| thin_course.course_code != course.course_code);
+    let mut new_selected = selected.clone();
+    new_selected
+        .courses
+        .retain(|thin_course, _| thin_course.course_code != course.course_code);
 
-        user_state.make_cookie(term)
-    });
+    let jar = CookieJar::new().add(new_selected.make_cookie(term));
+
+    let courses = state.courses(term, &new_selected.thin_courses())?;
 
     Ok((
         jar,
         html! {
             (calendar_view_container(true))
-            (courses_container(true))
+            (courses_container(true, term, &courses, &new_selected))
         },
     ))
 }
+
+// HANDLER FOR COURSE SECTIONS
+// because Axum doesn't support duplicated fields in form data...
+#[derive(Debug, PartialEq)]
+struct SectionQuery(ThinCourse, Vec<ThinSection>);
+
+impl TryFrom<&RawForm> for SectionQuery {
```

Comment:
Please use [this `Form` extractor instead](https://docs.rs/axum-extra/latest/axum_extra/extract/struct.Form.html#differences-from-axumextractform), it's supported alongside the official axum project.